app/grandchallenge/evaluation/views/api.py

```diff
@@ -23,3 +46,118 @@ class EvaluationViewSet(ReadOnlyModelViewSet):
+        evaluation.claimed_by = request.user
+        evaluation.started_at = now()
+        evaluation.save()
+
+        serializer = ExternalEvaluationSerializer(
+            instance=evaluation, context={"request": request}
+        )
+        return Response(serializer.data)
+
+    @action(detail=False, methods=["GET"])
+    def claimable_evaluations(self, request, *args, **kwargs):
+        queryset = get_objects_for_user(
+            request.user, "evaluation.claim_evaluation"
+        ).filter(status=Evaluation.PENDING)
+        serializer = ExternalEvaluationSerializer(
+            queryset, many=True, context={"request": request}
+        )
+        return Response(serializer.data)
+
+    @action(
+        detail=True,
+        methods=["PATCH"],
+        permission_classes=[CanClaimEvaluation],
+        serializer_class=ExternalEvaluationUpdateSerializer,
+    )
+    def update_external_evaluation(self, request, *args, **kwargs):
+        evaluation = self.get_object()
+        if evaluation.status != Evaluation.CLAIMED:
+            return Response(
+                {
+                    "status": "You need to claim an evaluation before you can update it."
+                },
+                status=400,
+            )
+
+        if request.user != evaluation.claimed_by:
+            return Response(
+                {
+                    "status": "You do not have permission to update this evaluation."
+                },
+                status=403,
+            )
+
+        if (
+            evaluation.started_at - now()
+        ).seconds > settings.EXTERNAL_EVALUATION_TIMEOUT_IN_SECONDS:
+            evaluation.status = Evaluation.CANCELLED
+            evaluation.error_message = "External evaluation timed out."
+            evaluation.save()
+            return Response(
+                {
+                    "status": "You can only update an evaluation within 24 hours."
+                },
+                status=400,
+            )
+
+        serialized_evaluation = ExternalEvaluationUpdateSerializer(
+            instance=evaluation,
+            data=request.data,
+            context={"request": request},
+        )
+        if not serialized_evaluation.is_valid():
+            raise DRFValidationError(serialized_evaluation.errors)
+
+        status = request.data.get("status", None)
+
+        if status == Evaluation.SUCCESS:
+            metrics = request.data.pop("metrics", None)
+            interface = ComponentInterface.objects.get(
+                slug="metrics-json-file"
+            )
+
+            civ = ComponentInterfaceValue(interface=interface, value=metrics)
+            try:
+                civ.full_clean()
+                civ.save()
+                evaluation.outputs.add(civ)
+                evaluation.status = Evaluation.SUCCESS
+            except ValidationError as e:
+                evaluation.status = Evaluation.FAILURE
+                evaluation.error_message = str(e)
```

Comment:
Does it make sense to put `ValidationError`s on the evaluation? Those are for the submitter, whereas the `ValidationError`s are for the evaluator. I don't think it makes sense to save the object and mark as Failed here as the evaluator could potentially fix the issue, and again it would be a case of returning 4xx but changing the object.

Reply:
Don't we normally put `ValidationErrors` on the evaluation as well? I do agree though that it doesn't make sense to share evaluation failure with the submitter, but it's not clear to me what to do instead. Marking it as cancelled doesn't make sense either, does it? Also the evaluator could not fix it then since cancelled evaluations can't be re-claimed. If we want the latter, we should leave the status at pending and just send a notification to the evaluator.

Reply:
> Don't we normally put ValidationErrors on the evaluation as well? 

Only `ComponentError`s should be put there. `ComponentError` means this is information for the participant and the error message has been reviewed. We should not be putting `ValidationError`s there unless they are for the participant. We do not need to communicate with the participant here as the `ValidationError` comes from a call from the evaluator, who needs to fix the call they make.

The evaluator should be a user that has an unusable password so does not log in to the web interface.

The evaluator will see the error when they do:

```python
response = requests.post(...)
response.raise_for_status()
```

They will then hopefully get notified by their internal systems. They can then fix it and re-play the `post` command. They would not need to claim the job again. If they do not do this then we will mark it as timed out.

Reply:
Ok, so we just raise the error, that's it.

Reply:
I think so! Double check the DRF docs.